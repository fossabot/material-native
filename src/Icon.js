'use strict';
import React, {PureComponent} from 'react';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import withMaterialTheme from './styles/withMaterialTheme';

export function createIconComponent(VectorIcon, displayName='Icon') {
	class Icon extends PureComponent {
		static defaultProps = {
			inactive: false,
			size: 24,
		};

		render() {
			const {materialTheme, inactive, size, color, ...props} = this.props;
			const themeColor = inactive ? materialTheme.icon.inactiveColor : materialTheme.icon.activeColor;

			return (
				<VectorIcon
					{...props}
					color={color || themeColor}
					size={size} />
			);
		}
	}
	Icon.displayName = displayName;

	return withMaterialTheme(Icon);
}

export const Icon = createIconComponent(MaterialIcons, 'Icon');
export const CommunityIcon = createIconComponent(MaterialCommunityIcons, 'CommunityIcon');
