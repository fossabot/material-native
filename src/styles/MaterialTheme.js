'use strict';
import * as colors from './colors';
import shades from './shades';
const F = Object.freeze;

function normalizeTheme(theme) {
	return theme === 'dark' ? 'dark' : 'light';
}

function inverse(theme) {
	return theme === 'dark' ? 'light' : 'dark';
}

const themeTones = F({
	light: F({
		statusBar: colors.grey300,
		barStyle: 'dark-content',
		appBar: colors.grey100,
		background: colors.grey50,
		container: colors.white,
	}),
	dark: F({
		statusBar: colors.black,
		barStyle: 'light-content',
		appBar: colors.grey900,
		background: '#303030',
		container: colors.grey800,
	}),
});

export default class MaterialTheme {
	static defaultTheme = new MaterialTheme();

	constructor(settings) {
		this.settings = settings || {};
		// @todo recursively freeze settings
		this._applyDefaults();
		this._buildTheme();
		this._freeze();
	}

	_applyDefaults() {
		this.settings.theme = normalizeTheme(this.settings.theme);
	}

	_buildTheme() {
		const {theme} = this.settings;
		const inverseTheme = inverse(theme);

		this.themeTone = theme;
		this.inverseThemeTone = inverseTheme;

		this.palette = {
			primary: this.settings.primary || colors.indigo500,
			darkPrimary: this.settings.darkPrimary || colors.indigo700,
			accent: this.settings.accent || colors.pinkA200,
			background: themeTones[theme].background,
			container: themeTones[theme].container,
		};

		this.defaultToolbar = Object.assign({
			theme,
			background: themeTones[theme].appBar,
		}, this.settings.defaultToolbar || {});

		// @todo Automatically pick the right appBar theme based on the appBar background
		//       Many solid appBar colors should implicitly turn the theme within the appBar to dark
		const appBarTheme = normalizeTheme((this.settings.appBar && this.settings.appBar.theme) || this.defaultToolbar.theme);
		this.appBar = Object.assign({}, this.defaultToolbar, {
			theme: appBarTheme,
			background: this.palette.primary || this.defaultToolbar.background,
		}, this.settings.appBar || {});

		this.statusBar = Object.assign({
			background: this.palette.darkPrimary || themeTones[inverse(appBarTheme)].statusBar,
			translucentBackground: shades.dark.statusBar,
			barStyle: themeTones[inverse(appBarTheme)].barStyle,
		}, this.settings.statusBar || {});

		this.text = Object.assign({
			primaryColor: shades[inverseTheme].primaryText,
			secondaryColor: shades[inverseTheme].secondaryText,
			disabledColor: shades[inverseTheme].disabledText,
			hintColor: shades[inverseTheme].hintText,
			primaryOpacity: shades.opacity[inverseTheme].primaryText,
			secondaryOpacity: shades.opacity[inverseTheme].secondaryText,
			disabledOpacity: shades.opacity[inverseTheme].disabledText,
			hintOpacity: shades.opacity[inverseTheme].hintText,
		}, this.settings.text || {});

		this.icon = Object.assign({
			activeColor: shades[inverseTheme].activeIcon,
			inactiveColor: shades[inverseTheme].inactiveIcon,
			activeOpacity: shades.opacity[inverseTheme].activeIcon,
			inactiveOpacity: shades.opacity[inverseTheme].inactiveIcon,
		}, this.settings.icon || {});
	}

	_freeze() {
		Object.freeze(this); // @todo recursive
	}

	/**
	 * Create a new theme object that inherits settings from another
	 */
	extend(settings) {
		// @todo immutably merge `this.settings` with `settings` and pass that to `new MaterialTheme()`
	}
}

export function getMaterialTheme(settings) {
	return new MaterialTheme(settings);
}
