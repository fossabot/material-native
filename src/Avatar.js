'use strict';
import React, {PureComponent} from 'react';
import {StyleSheet, View, Text} from 'react-native';
import withMaterialTheme from './styles/withMaterialTheme';
import * as colors from './styles/colors';
import * as typo from './styles/typo';
import {largePrimaryTextShade} from './styles/wcag';

class Avatar extends PureComponent {
	static defaultProps = {
		size: 40,
		color: colors.grey500,
	};

	render() {
		const {
			style,
			size,
			color,
			image,
			icon,
			text,
			textColor: textColorOverride,
		} = this.props;

		if ( typeof image === 'string' ) throw new TypeError('image must be a react element, not a string');
		if ( typeof icon === 'string' ) throw new TypeError('icon must be a react element, not a string');

		const borderRadius = size / 2;
		const textColor = textColorOverride || largePrimaryTextShade(color);

		let avatar;

		if ( image ) {
			avatar = React.cloneElement(image, {
				style: {
					width: size,
					height: size,
					borderRadius,
				},
			});
		} else if ( icon ) {
			avatar = React.cloneElement(icon, {
				color: textColor,
				size: 0.6 * size,
			});
		} else if ( text ) {
			// Standard text size is 18dp in a 40dp avatar, use that to calculate the text size for this avatar size
			const textHeight = (18/40) * size;
			// A font size of 18dp actually results in avatar text about 13db high, so calculate the font size with that adjustment
			const fontSize = (18/13) * textHeight;

			avatar = (
				<Text
					suppressHighlighting
					numberOfLines={1}
					style={[
						styles.text,
						{
							color: textColor,
							fontSize,
						}
					]}>
					{text}
				</Text>
			);
		}

		if ( !avatar ) {
			return null;
		}

		return (
			<View
				style={[
					style,
					styles.container,
					{
						width: size,
						height: size,
						borderRadius,
					},
					!image && {backgroundColor: color}
				]}>
				{avatar}
			</View>
		);
	}
}

export default withMaterialTheme(Avatar);

const styles = StyleSheet.create({
	container: {
		alignItems: 'center',
		justifyContent: 'center',
	},
	text: {
		...typo.fontLight,
		textAlign: 'center',
		textAlignVertical: 'center',
		includeFontPadding: false,
	},
});
