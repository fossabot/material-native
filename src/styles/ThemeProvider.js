'use strict';
import Observable from 'zen-observable';
import PropTypes from '../PropTypes';
import React, {PureComponent} from 'react';
import MaterialTheme from './MaterialTheme';

const defaultTheme = new MaterialTheme();

export default class ThemeProvider extends PureComponent {
	static contextTypes = {
		materialThemeObservable: PropTypes.Observable,
	};

	static childContextTypes = {
		materialThemeObservable: PropTypes.Observable.isRequired,
		memoizedMaterialStyles: PropTypes.func,
	};

	getChildContext() {
		return {
			materialThemeObservable: this.observable,
			memoizedMaterialStyles: this.memoizedMaterialStyles,
		};
	}

	componentWillMount() {
		// materialTheme
		if ( this.context.materialThemeObservable ) {
			this._parentSubscription = this.context.materialThemeObservable.subscribe({
				next: (theme) => {
					this.parentMaterialTheme = theme;
					this.handleThemeChange(this.props);
				},
				complete: () => {
					delete this._parentSubscription;
				},
			});
		} else {
			this.handleThemeChange(this.props);
		}
	}

	componentWillReceiveProps(nextProps) {
		// materialTheme
		this.handleThemeChange(nextProps);
	}

	componentWillUnmount() {
		if ( this._parentSubscription ) {
			this._parentSubscription.unsubscribe();
		}
	}

	// materialTheme
	observable = new Observable((observer) => {
		if ( this.materialTheme ) {
			observer.next(this.materialTheme);
		}

		this.observers.add(observer);

		return () => {
			this.observers.delete(observer);
		};
	});

	observers = new Set();

	materialTheme = undefined;
	parentMaterialTheme = undefined;

	// materialTheme
	handleThemeChange(props) {
		const baseTheme = props.theme || this.parentMaterialTheme || defaultTheme;

		// @todo Handle extend prop here

		if ( baseTheme !== this.materialTheme ) {
			this.materialTheme = baseTheme;
			this.observers.forEach((observer) => observer.next(this.materialTheme));
		}
	}

	// styleCache
	materialStyleCache = new Map();

	memoizedMaterialStyles = (materialTheme, mapThemeToStyles) => {
		let cache = this.materialStyleCache.get(mapThemeToStyles);
		if ( cache && cache.materialTheme === materialTheme ) {
			return cache.styles;
		}

		// @note If possible in the future, styles in `cache` can be unloaded at this point

		const cacheStyles = mapThemeToStyles(materialTheme);

		for ( const key in cacheStyles ) {
			Object.freeze(cacheStyles[key]);
		}
		Object.freeze(cacheStyles);

		cache = {
			materialTheme,
			styles: cacheStyles,
		};

		this.materialStyleCache.set(mapThemeToStyles, cache);

		return cache.styles;
	};

	// boilerplate
	render() {
		return React.Children.only(this.props.children);
	}
}
