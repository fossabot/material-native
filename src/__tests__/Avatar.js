'use strict';
import React from 'react';
import {Image} from 'react-native';
import MaterialTheme from '../styles/MaterialTheme';
import ThemeProvider from '../styles/ThemeProvider';
import Avatar from '../Avatar';
import {CommunityIcon} from '../Icon';

import ReactTestRenderer from 'react-test-renderer';
import ToJSX from 'react-element-to-jsx-string';

const theme = new MaterialTheme({});

/* eslint-disable react/jsx-key */
const snapshotTests = [
	<Avatar text='A' />,
	<Avatar text='A' color='#123456' />,
	<Avatar text='A' textColor='#123456' />,
	<Avatar icon={<CommunityIcon name='account' />} />,
	<Avatar icon={<CommunityIcon name='account' />} color='#123456' />,
	<Avatar image={<Image source={{uri: 'http://example.com/example.png'}} />} />,
	<Avatar size={100} text='A' />,
	<Avatar size={100} icon={<CommunityIcon name='account' />} />,
	<Avatar size={100} image={<Image source={{uri: 'http://example.com/example.png'}} />} />
];
/* eslint-enable react/jsx-key */

for ( const root of snapshotTests ) {
	test(ToJSX(root, {maxInlineAttributesLineLength: Infinity}), () => {
		const host = ReactTestRenderer.create(
			<ThemeProvider theme={theme}>
				{root}
			</ThemeProvider>
		);
		expect(host.toJSON()).toMatchSnapshot();
	});
}
