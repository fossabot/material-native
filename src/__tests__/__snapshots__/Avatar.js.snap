// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withMaterialTheme(Avatar) color="#123456" icon={<withMaterialTheme(CommunityIcon) name="account" />} /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      Object {
        "backgroundColor": "#123456",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "rgba(255, 255, 255, 1)",
          "fontSize": 24,
        },
        Array [
          undefined,
          undefined,
        ],
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) color="#123456" text="A" /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      Object {
        "backgroundColor": "#123456",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontWeight": "300",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "rgba(255, 255, 255, 1)",
          "fontSize": 24.923076923076923,
        },
      ]
    }
    suppressHighlighting={true}
  >
    A
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) icon={<withMaterialTheme(CommunityIcon) name="account" />} /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      Object {
        "backgroundColor": "#9E9E9E",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontSize": 24,
        },
        Array [
          undefined,
          undefined,
        ],
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) icon={<withMaterialTheme(CommunityIcon) name="account" />} size={100} /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      },
      Object {
        "backgroundColor": "#9E9E9E",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontSize": 60,
        },
        Array [
          undefined,
          undefined,
        ],
        Object {
          "fontFamily": "Material Design Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) image={<Image source={{uri: 'http://example.com/example.png'}} />} /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      false,
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "http://example.com/example.png",
      }
    }
    style={
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      }
    }
  />
</View>
`;

exports[`<withMaterialTheme(Avatar) image={<Image source={{uri: 'http://example.com/example.png'}} />} size={100} /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      },
      false,
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "http://example.com/example.png",
      }
    }
    style={
      Object {
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      }
    }
  />
</View>
`;

exports[`<withMaterialTheme(Avatar) size={100} text="A" /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 50,
        "height": 100,
        "width": 100,
      },
      Object {
        "backgroundColor": "#9E9E9E",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontWeight": "300",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontSize": 62.30769230769231,
        },
      ]
    }
    suppressHighlighting={true}
  >
    A
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) text="A" /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      Object {
        "backgroundColor": "#9E9E9E",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontWeight": "300",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontSize": 24.923076923076923,
        },
      ]
    }
    suppressHighlighting={true}
  >
    A
  </Text>
</View>
`;

exports[`<withMaterialTheme(Avatar) text="A" textColor="#123456" /> 1`] = `
<View
  style={
    Array [
      undefined,
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      Object {
        "borderRadius": 20,
        "height": 40,
        "width": 40,
      },
      Object {
        "backgroundColor": "#9E9E9E",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontWeight": "300",
          "includeFontPadding": false,
          "textAlign": "center",
          "textAlignVertical": "center",
        },
        Object {
          "color": "#123456",
          "fontSize": 24.923076923076923,
        },
      ]
    }
    suppressHighlighting={true}
  >
    A
  </Text>
</View>
`;
